Doc/c-api/typeobj.rst

```diff
@@ -2094,32 +2227,141 @@ and :c:data:`PyType_Type` effectively act as defaults.)
+      beginning of :c:member:`!tp_dealloc` to guarantee that the object is
+      always finalized before destruction.
+
+   .. warning::
+
+      The :c:member:`~PyTypeObject.tp_finalize` function can be called from any
+      thread, although the :term:`GIL` will be held unless disabled (see
+      :ref:`whatsnew313-free-threaded-cpython`).
+
+   .. warning::
+
+      The :c:member:`!tp_finalize` function can be called during shutdown,
+      after some global variables have been deleted.  See the documentation of
+      the :meth:`~object.__del__` method for details.
+
+   When Python finalizes an object, it behaves like the following algorithm:
+
+   #. Python might mark the object as *finalized*.  Currently :term:`CPython`
+      always marks objects whose type supports garbage collection (i.e., the
+      :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set in
+      :c:member:`~PyTypeObject.tp_flags`) and never marks other types of
+      objects; this might change in a future version.
+   #. If the object is not marked as *finalized* and its
+      :c:member:`!tp_finalize` finalizer function is non-``NULL``, the
+      finalizer function is called.
+   #. If the finalizer function was called and the finalizer made the object
+      reachable (i.e., there is a reference to the object and it is not a
+      member of a :term:`cyclic isolate`), then the finalizer is said to have
+      *resurrected* the object.  It is unspecified whether the finalizer can
+      also resurrect the object by adding a new reference to the object that
+      does not make it reachable, i.e., the object is (still) a member of a
+      cyclic isolate.
+   #. If the finalizer resurrected the object, the object's pending destruction
+      is canceled and the object's *finalized* mark might be cleared if
+      present.  Currently :term:`CPython` never clears the *finalized* mark;
+      this might change in a future version.
+
+   *Automatic finalization* refers to any finalization performed by Python
+   except via calls to :c:func:`PyObject_CallFinalizer` or
+   :c:func:`PyObject_CallFinalizerFromDealloc`.  No guarantees are made about
+   when, if, or how often an object is automatically finalized, except:
+
+   * Python will not automatically finalize an object if it is reachable, i.e.,
+     there is a reference to it and it is not a member of a :term:`cyclic
+     isolate`.
+   * Python will not automatically finalize an object if finalizing it would
+     not mark the object as *finalized*.  In :term:`CPython` this currently
+     applies to objects whose type does not support garbage collection, i.e.,
+     the :c:macro:`Py_TPFLAGS_HAVE_GC` flag is not set.  Such objects can still
+     be manually finalized by calling :c:func:`PyObject_CallFinalizer` or
+     :c:func:`PyObject_CallFinalizerFromDealloc`.
+   * Python will not automatically finalize any two members of a :term:`cyclic
+     isolate` concurrently.
```

Comment:
I think we're over-documenting for a type slot. Instead of repeating the details of the finalization process, we should just link to the other page with the rest of the garbage collector information.

Reply:
That was my intention—just document the relationship between the slots in this page, but leave the specifics to the slot-specific documentation.  I added a paragraph at the top to clarify.